syscall/devices/hostepoll/hostepoll.c

```diff
@@ -457,26 +465,32 @@ static int _epoll_wait(
         if (retval > maxevents)
             OE_RAISE_ERRNO(OE_EINVAL);
 
+        locked = true;
+        oe_mutex_lock(&epoll->lock);
+
         for (int i = 0; i < retval; i++)
         {
-            struct oe_epoll_event* event = &events[i];
-            const mapping_t* mapping;
+            struct oe_epoll_event* const event = &events[i];
+            const mapping_t* const mapping = _map_find(epoll, event->data.fd);
 
-            oe_spin_lock(&epoll->lock);
+            if (mapping)
+                event->data.u64 = mapping->event.data.u64;
+            else
             {
-                if ((mapping = _map_find(epoll, event->data.fd)))
-                    event->data.u64 = mapping->event.data.u64;
+                // fd has been deleted between the return of epoll_wait and the
+                // acquisition of the lock.
+                --retval;
+                memmove(
+                    event, event + 1, (size_t)(retval - i) * sizeof(*event));
             }
-            oe_spin_unlock(&epoll->lock);
-
-            if (!mapping)
-                OE_RAISE_ERRNO(OE_ENOENT);
         }
     }
 
     ret = (int)retval;
 
 done:
+    if (locked)
```

Comment:
Thanks for this fix. I am wondering why you decided to move from spinlock to mutex.

Reply:
I changed it because the lock is now acquired before calling oe_syscall_epoll_ctl_ocall(). That seemed to be a too expensive operation to me. However, I didn't do any measuring to justify this change. What do you think?

Reply:
Thanks!

Reply:
I do agree that an ocall (and subsequent kernel call) seems like too expensive of an operation to hold a spinlock. I think this is the right approach.

@mikbras can you take a look at this as well? I'm a little worried about holding a lock and then making an ocall since threads can be re-entrant and cause deadlock. It makes me nervous but in this case epoll doesn't try to re-enter the enclave for any reason so my fear is probably unwarranted.